### DISCUSSION TITLE
Provide function to serialize a State

### DISCUSSION THREAD

**Discussion by maexrakete** [statelyai/xstate#2064] [Ideas]
Hey!

right now the suggested way of serializing and persisting the state is using JSON.stringify and JSON.parse according to the docs: https://xstate.js.org/docs/guides/states.html#persisting-state

This has following downsides:

1. A state serialized in that way loses all its activities and actions, since functions can't be serialized. The resulting type is by definition no longer compatible with the `StateConfig` type
2. It's not transparent that the information-loss mentioned in 1. happens and this might lead to errors when loading a serialized state whose defined activities or actions are lost, 
3. `JSON.parse(JSON.stringify(state))` is really really ugly

Following suggestion could mitigate that problem:

0. **Introduce a dedicated type for a serialized state (eg. `SerializedState`)**
1. **Introduce a function to dehydrate a State into a `SerializedState`.**
I don't know enough about XState to make a solid suggestion, but as an educated guess it could include the current value, context and the history. 
2. **Provide a way to rehydrate a `SerializedState` in a concrete `State`** 
This could be a deep-merge of the serialized state onto the inital state, or something along that lines. For ergonomic reasons that could be a function directly on the StateMachine

Let me know if you have any ideas, improvements or doubts about that idea. I'd be happy to provide a PR, when we've specified how that feature could work.

Cheers!

**Comment by awreccan** [comment]
+1 requesting a similar tool here: https://github.com/davidkpiano/xstate/discussions/2160
